File: app/src/main/java/me/zhanghai/android/files/provider/common/AccessModes.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common;

import java.util.Objects;

import java8.nio.file.AccessMode;

public class AccessModes {

    private final boolean mRead;
    private final boolean mWrite;
    private final boolean mExecute;

    public AccessModes(boolean read, boolean write, boolean execute) {
        mRead = read;
        mWrite = write;
        mExecute = execute;
    }

    public boolean hasRead() {
        return mRead;
    }

    public boolean hasWrite() {
        return mWrite;
    }

    public boolean hasExecute() {
        return mExecute;
    }

    public static AccessModes fromArray(AccessMode[] modes) {
        boolean read = false;
        boolean write = false;
        boolean execute = false;
        for (AccessMode mode : modes) {
            Objects.requireNonNull(mode);
            switch (mode) {
                case READ:
                    read = true;
                    break;
                case WRITE:
                    write = true;
                    break;
                case EXECUTE:
                    execute = true;
                    break;
                default:
                    throw new UnsupportedOperationException(mode.toString());
            }
        }
        return new AccessModes(read, write, execute);
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/common/AccessModes.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.common

import java8.nio.file.AccessMode

class AccessModes(val read: Boolean, val write: Boolean, val execute: Boolean)

fun Array<out AccessMode>.toAccessModes(): AccessModes {
    var read = false
    var write = false
    var execute = false
    for (mode in this) {
        when (mode) {
            AccessMode.READ -> read = true
            AccessMode.WRITE -> write = true
            AccessMode.EXECUTE -> execute = true
            else -> throw UnsupportedOperationException(mode.toString())
        }
    }
    return AccessModes(read, write, execute)
}
